wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift

```diff
@@ -0,0 +1,131 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public typealias OAuthBlock = (_ idP: URL,
+                               _ clientID: String,
+                               _ keyauth: String,
+                               _ acmeAud: String) async throws -> (String, String)
+
+public protocol EnrollE2eICertificateUseCaseInterface {
+
+    func invoke(authenticate: OAuthBlock) async throws
+
+}
+
+/// This class provides an interface to issue an E2EI certificate.
+public final class EnrollE2eICertificateUseCase: EnrollE2eICertificateUseCaseInterface {
+
+    // MARK: - Types
+
+    enum Failure: Error {
+        case missingIdentityProvider
+        case missingClientId
+        case failedToDecodeCertificate
+        case failedToEnrollCertificate(_ underlyingError: Error)
+    }
+
+    // MARK: - Properties
+
+    private let logger = WireLogger.e2ei
+    private let e2eiRepository: E2eIRepositoryInterface
+    private let context: NSManagedObjectContext
+
+    // MARK: - Life cycle
+
+    public init(e2eiRepository: E2eIRepositoryInterface,
+                context: NSManagedObjectContext) {
+        self.e2eiRepository = e2eiRepository
+        self.context = context
+    }
+
+    public func invoke(authenticate: OAuthBlock) async throws {
+        do {
```

Comment:
suggestion: link here the relevant documentation in confluence

Reply:
https://github.com/wireapp/wire-ios/pull/1029